### ISSUE TITLE
Phase 48.1: ElasticWeightConsolidator — EWC, Fisher Information & Parameter Importance Estimation

### ISSUE THREAD

**Issue by web3guru888** [web3guru888/asi-build#938]
### Overview & Motivation

Catastrophic forgetting occurs when gradient updates for new tasks overwrite parameters critical for previously learned tasks. **Elastic Weight Consolidation (EWC)** addresses this by selectively constraining parameter updates based on their importance to prior tasks, measured via the Fisher information matrix.

The ElasticWeightConsolidator implements EWC and related regularization-based continual learning methods (Synaptic Intelligence, Memory Aware Synapses) that protect important parameters while allowing plasticity for new learning.

### Architecture Components

```
ElasticWeightConsolidator
├── FisherInformationComputer
│   ├── DiagonalFisherEstimator      # Diagonal approximation of Fisher matrix
│   ├── KroneckerFisherEstimator     # KFAC-style block-diagonal approximation
│   ├── EmpiricalFisherCalculator    # Sample-based Fisher estimation
│   └── OnlineFisherAccumulator      # Running Fisher across tasks (Online EWC)
├── ParameterImportanceEstimator
│   ├── EWCImportance                # Fisher diagonal as importance
│   ├── SIImportance                 # Path integral importance (Zenke et al.)
│   ├── MASImportance                # Output sensitivity importance (Aljundi et al.)
│   └── HybridImportance             # Ensemble of multiple importance signals
├── RegularizationEngine
│   ├── QuadraticPenaltyComputer     # L_ewc = λ/2 Σ F_i (θ_i - θ*_i)²
│   ├── OnlineConsolidator           # Running parameter of Schwarz et al. (2018)
│   ├── TaskSpecificRegularizer      # Per-task penalty terms
│   └── AdaptiveLambdaScheduler      # Dynamic regularization strength
├── ParameterSnapshotManager
│   ├── TaskCheckpointer             # Store θ* after each task
│   ├── ImportanceArchive            # Store F_i per task
│   ├── SnapshotCompressor           # Memory-efficient storage
│   └── MergingStrategy              # Merge importance across tasks
└── DiagnosticsEngine
    ├── ForgettingMetrics             # Backward transfer measurement
    ├── PlasticityMetrics             # Forward transfer / learning ability
    ├── ImportanceVisualization        # Parameter importance heatmaps
    └── StabilityPlasticityTracker    # Balance monitoring
```

### Key Algorithms & Data Structures

**EWC Loss Function:**
```
L_total(θ) = L_new(θ) + (λ/2) Σ_i F_i · (θ_i - θ*_i)²
```
where F_i is the diagonal Fisher information for parameter i, θ*_i is the optimal value after previous tasks, and λ controls regularization strength.

**Fisher Information Computation:**
```python
class DiagonalFisherEstimator:
    def compute(self, model, dataloader, num_samples=200):
        fisher = {n: torch.zeros_like(p) for n, p in model.named_parameters()}
        for x, y in itertools.islice(dataloader, num_samples):
            model.zero_grad()
            log_prob = F.log_softmax(model(x), dim=1)
            y_sampled = Categorical(logits=log_prob).sample()
            loss = F.nll_loss(log_prob, y_sampled)
            loss.backward()
            for n, p in model.named_parameters():
                fisher[n] += p.grad.data ** 2 / num_samples
        return fisher
```

**Synaptic Intelligence (SI):**
```python
class SIImportance:
    def update_running_sum(self, param_name, grad, delta_param):
        self.omega[param_name] += (-grad * delta_param).data

    def compute_importance(self, param_name, old_param, new_param, xi=0.1):
        delta = (new_param - old_param) ** 2 + xi
        return self.omega[param_name] / delta
```

**Online EWC (Schwarz et al., 2018):**
```python
class OnlineFisherAccumulator:
    def accumulate(self, new_fisher, gamma=0.95):
        for name in self.running_fisher:
            self.running_fisher[name] = gamma * self.running_fisher[name] + new_fisher[name]
```

### API Surface

```python
class ElasticWeightConsolidator:
    def __init__(self, config: EWCConfig): ...
    def register_model(self, model: nn.Module) -> None: ...
    def compute_fisher(self, dataloader: DataLoader, method: str = "diagonal") -> FisherInfo: ...
    def consolidate_task(self, task_id: str) -> ConsolidationResult: ...
    def compute_penalty(self, model: nn.Module) -> torch.Tensor: ...
    def get_importance_map(self, task_id: str = None) -> Dict[str, Tensor]: ...
    def get_forgetting_metrics(self, eval_results: Dict) -> ForgettingReport: ...
    def merge_importances(self, strategy: str = "online") -> MergedImportance: ...
    def export_state(self) -> ConsolidatorState: ...
```

### Testing Strategy

- Unit tests for Fisher computation accuracy (compare diagonal vs full on small networks)
- Verification that EWC penalty gradient is correct via finite differences
- Integration test: 2-task sequential learning showing reduced forgetting vs naive fine-tuning
- Benchmark on Split-MNIST and Permuted-MNIST
- Memory overhead tests for importance storage
- Online EWC vs standard EWC comparison

### References

1. Kirkpatrick et al. (2017) — *Overcoming catastrophic forgetting in neural networks*
2. Zenke et al. (2017) — *Continual Learning through Synaptic Intelligence*
3. Aljundi et al. (2018) — *Memory Aware Synapses*
4. Schwarz et al. (2018) — *Progress & Compress: A scalable framework for continual learning*
5. Martens (2020) — *New Insights and Perspectives on the Natural Gradient Method*
6. De Lange et al. (2021) — *A continual learning survey*

Part of **Phase 48: Continual Learning & Catastrophic Forgetting Prevention**
Discussion: #937
